dbms/src/Storages/DeltaMerge/VersionChain/VersionChain.cpp

```diff
@@ -0,0 +1,383 @@
+    const UInt32 delta_rows = delta.getRows();
+    const UInt32 delta_delete_ranges = delta.getDeletes();
+    std::lock_guard lock(mtx); // Not run concurrently. Because it can reuse the result of the previous replay.
+    if (delta_rows + delta_delete_ranges <= replayed_rows_and_deletes)
+    {
+        RUNTIME_CHECK(base_versions->size() >= delta_rows, base_versions->size(), delta_rows);
+        return base_versions;
+    }
+
+    try
+    {
+        return replaySnapshotImpl(dm_context, snapshot);
+    }
+    catch (DB::Exception & e)
+    {
+        LOG_ERROR(snapshot.log, "errmsg: {}", e.message());
+        resetImpl();
+        throw;
+    }
+    catch (std::exception & e)
+    {
+        LOG_ERROR(snapshot.log, "errmsg: {}", e.what());
+        resetImpl();
+        throw;
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+        resetImpl();
+        throw;
+    }
+}
+
+template <ExtraHandleType HandleType>
+std::shared_ptr<const std::vector<RowID>> VersionChain<HandleType>::replaySnapshotImpl(
+    const DMContext & dm_context,
+    const SegmentSnapshot & snapshot)
+{
+    Stopwatch sw_total;
+    const auto & stable = *(snapshot.stable);
+    const UInt32 stable_rows = stable.getDMFilesRows();
+    const auto & delta = *(snapshot.delta);
+    const UInt32 delta_rows = delta.getRows();
+    const UInt32 delta_delete_ranges = delta.getDeletes();
+
+    if (dmfile_or_delete_range_list.empty())
+    {
+        // In theory, we can support stable composed of multiple disjoint dmfiles in stable.
+        // But it is not necessary for now. For simplicity, assume stable always has one DMFile.
+        const auto & dmfiles = snapshot.stable->getDMFiles();
+        RUNTIME_CHECK(dmfiles.size() == 1, dmfiles.size());
+        dmfile_or_delete_range_list.push_back(
+            DMFileHandleIndex<HandleType>{dm_context, dmfiles[0], /*start_row_id*/ 0, /*rowkey_range*/ std::nullopt});
+    }
+
+    // base_versions may be shared for read, so copy for write here.
+    base_versions = std::make_shared<std::vector<RowID>>(*base_versions);
+    const auto cfs = delta.getColumnFiles();
+    const auto & data_provider = delta.getDataProvider();
+
+    UInt32 skipped_rows_and_deletes = 0;
+    auto pos = cfs.begin();
+    for (; pos != cfs.end(); ++pos)
+    {
+        auto skip_n = (*pos)->isDeleteRange() ? (*pos)->getDeletes() : (*pos)->getRows();
+        if (skip_n + skipped_rows_and_deletes > replayed_rows_and_deletes)
+            break;
+        skipped_rows_and_deletes += skip_n;
+    }
+    // `pos` points to the first ColumnFile that has records not been replayed.
+    // `offset` points to the first records that has not been replayed in `pos`.
+    auto offset = replayed_rows_and_deletes - skipped_rows_and_deletes;
+    // Only ColumnFileInMemory or ColumnFileTiny can be half replayed.
+    RUNTIME_CHECK(offset == 0 || (*pos)->isInMemoryFile() || (*pos)->isTinyFile(), offset, (*pos)->toString());
```

Comment:
```suggestion
    RUNTIME_CHECK(pos != cfs.end(), skipped_rows_and_deletes, replayed_rows_and_deletes);
    RUNTIME_CHECK(offset == 0 || (*pos)->isInMemoryFile() || (*pos)->isTinyFile(), offset, (*pos)->toString());
```

In case we run into cases that break the assumption and cause crash in check